### PR TITLE
CM-APP-53 | Added lastUpdated column, sorting by last updated, filtering GRANTED consents

### PR DESCRIPTION
**Pull request by shwetadongare**

### CODE REVIEW

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java [hunk 1]
```diff
@@ -47,7 +50,7 @@
     static {
         String s = "SELECT request_id, status, details, date_created, date_modified FROM consent_request " +
                 "where ";
-        SELECT_CONSENT_DETAILS_FOR_PATIENT = s + "patient_id=$1 and (status=$4 OR $4 IS NULL) " +
+        SELECT_CONSENT_DETAILS_FOR_PATIENT = s + "patient_id=$1 and status!=$5 and (status=$4 OR $4 IS NULL) " +
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java:53]: why is this required? status != GRANTED? If you are filtering by status=$4, would it not get covered? 

also, is there a case ever that status will be null?

Comment (shwetadongare) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java:53]: Q: why is this required? status != GRANTED? If you are filtering by status=$4, would it not get covered?
A: This is required because we have to filter the requests by GRANTED and we have an ALL request param which return all status in order to filter it from ALL a separate clause is required. 

Q: is there a case ever that status will be null?
A: Yes, when we get 'ALL' as request param we make the status null from service after a check, basically null is ALL statuses.

Comment (angshu) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java:53]: Q: how do you get only GRANTED requests then? 
Q? In the DB, will there be ever a case that the status is null? If so, why - even when the request is created - then it should say that the request is PENDING or something right?

Comment (shwetadongare) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java:53]: Q: how do you get only GRANTED requests then?
A: I believe we have a separate call for that - getConsentsByStatus

Q: In the DB, will there be ever a case that the status is null? If so, why - even when the request is created - then it should say that the request is PENDING or something right?
A: From what I could figure, when we insert a request into DB we insert as REQUESTED and then the status keeps changing when we GRANT or REVOKE or it Expires. So no, there isn't a case where in DB the status will be null.

Comment (angshu) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java:53]: right .. so we don't need a null check. and please check if the getConsentsByStatus is redundant?

Comment (shwetadongare) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java:53]: Q: is there a case ever that status will be null?
A: Yes, when we get 'ALL' as request param we make the status null from service after a check, basically null is ALL statuses.
Like so, https://github.com/ProjectEKA/consent-manager/blob/CM-APP-43/src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java#L176

Also getConsentsByStatus is the only way to get GRANTED so no it's not redundant.

Comment (angshu) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java:53]: ok. my question was whether the status in DB ever will be null. If you get STATUS as empty or null as part of request, you can just not include the status filter right? Unless you dont want to write another query at all.

Comment (angshu) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java:53]: and we dont need the null check in the query - right?

File: src/main/java/in/projecteka/consentmanager/consent/ConsentArtefactRepository.java [hunk 2]
```diff
@@ -166,7 +168,10 @@
         for (Row result : results) {
             artefacts.add(getConsentArtefactRepresentation(result));
         }
-        return artefacts;
+        return artefacts.stream().
+                sorted(Comparator.comparing(c ->
+                        c.getConsentDetail().getLastUpdated(), Comparator.reverseOrder()))
+                .collect(Collectors.toList());
```
Comment (dheeraj-p) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentArtefactRepository.java:174]: We can sort it in the query itself on basis of date_modified field. If I am not wrong.

File: src/main/java/in/projecteka/consentmanager/consent/ConsentArtefactRepository.java [hunk 3]
```diff
@@ -268,6 +273,7 @@ private Date convertToDate(LocalDateTime timestamp) {
     private ConsentArtefactRepresentation getConsentArtefactRepresentation(Row row) {
         ConsentArtefact consentArtefact = to(row.getValue(CONSENT_ARTEFACT).toString(),
                 ConsentArtefact.class);
+        consentArtefact.setLastUpdated(convertToDate(row.getLocalDateTime(DATE_MODIFIED)));
```
Comment (dheeraj-p) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentArtefactRepository.java:276]: We can set in the builder chain instead of separately calling a setter method.

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java [hunk 4]
```diff
@@ -103,7 +106,9 @@ public ConsentRequestRepository(PgPool dbClient) {
             ConsentRequestDetail aDetail = mapToConsentRequestDetail(result);
             requestList.add(aDetail);
         }
-        return requestList;
+        return requestList.stream().
+                sorted(Comparator.comparing(ConsentRequestDetail::getLastUpdated).reversed())
+                .collect(Collectors.toList());
```
Comment (dheeraj-p) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentRequestRepository.java:111]: Here also we can sort it in the query itself.
